File: algoritmos-y-estructuras-de-datos-2/algorithms/sorting/countingSort.hs
main :: IO ()
main = do
  print (accumulate [1, 0, 0, 0, 6])


isSorted :: Ord a => [a] -> Bool
isSorted []  = True
isSorted [x] = True
isSorted (x:y:ys)
  | x <= y    = isSorted (y:ys)
  | otherwise = False
  
  
incAt :: Int -> [Int] -> [Int]
incAt _ []     = []
incAt 0 (y:ys) = (y+1) : ys
incAt n (y:ys) = y : incAt (n-1) ys

decAt :: Int -> [Int] -> [Int]
decAt _ []     = []
decAt 0 (y:ys) = (y-1) : ys
decAt n (y:ys) = y : decAt (n-1) ys

counter :: Int -> [Int]
counter k = take k (repeat 0)

count :: [Int] -> [Int] -> [Int]
count [] ys     = ys
count (x:xs) ys =  count xs (incAt x ys)

accumulate :: [Int] -> [Int]
accumulate = scanl1 (+)

countingSort :: Ord a => [a] -> Int -> [a]
countingSort [] _  = []
countingSort [x] _ = [x]
countingSort xs k  = fuck haskell
  
example = [1, 0 , 3, 4, 5, -3, 20, 45, 2]

File: algoritmos-y-estructuras-de-datos-2/algorithms/sorting/countingSort.js
function countingSort(A, B, k) {
    let C = [];
    for (let i = 0; i < k; i++) {
        C[i] = 0;
    }
    // Count
    for (let j = 0; j < A.length; j++) {
        C[A[j]] = C[A[j]] + 1;
    }
    // Accumulate
    for (let i = 1; i < k; i++) {
        C[i] = C[i] + C[i-1];
    }
    // Sort
    for (let i = A.length - 1; i >= 0; i--) {
        B[C[A[i]] - 1] = A[i];
        C[A[i]] = C[A[i]] - 1;
    }
    return B;
}


let A = [2, 0, 2, 3, 5, 9];
let B = [];

console.log(countingSort(A, B, 10));
